--- old/src/cpu/ppc/vm/assembler_ppc.hpp	2017-09-29 18:22:30.064712848 -0300
+++ new/src/cpu/ppc/vm/assembler_ppc.hpp	2017-09-29 18:22:29.976710758 -0300
@@ -1308,6 +1308,7 @@
   inline void li(   Register d, int si16);
   inline void lis(  Register d, int si16);
   inline void addir(Register d, int si16, Register a);
+  inline void subi( Register d, Register a, int si16);
 
   static bool is_addi(int x) {
      return ADDI_OPCODE == (x & ADDI_OPCODE_MASK);
