--- old/src/cpu/ppc/vm/stubGenerator_ppc.cpp	2017-09-27 11:38:57.766128589 -0300
+++ new/src/cpu/ppc/vm/stubGenerator_ppc.cpp	2017-09-27 11:38:57.686126809 -0300
@@ -2524,6 +2524,14 @@
       StubRoutines::_aescrypt_decryptBlock = generate_aescrypt_decryptBlock();
     }
 
+    if (UseMontgomeryMultiplyIntrinsic) {
+      StubRoutines::_montgomeryMultiply
+        = CAST_FROM_FN_PTR(address, SharedRuntime::montgomery_multiply);
+    }
+    if (UseMontgomerySquareIntrinsic) {
+      StubRoutines::_montgomerySquare
+        = CAST_FROM_FN_PTR(address, SharedRuntime::montgomery_square);
+    }
   }
 
  public:
