--- old/src/cpu/ppc/vm/vm_version_ppc.cpp	2017-09-27 11:38:58.806151729 -0300
+++ new/src/cpu/ppc/vm/vm_version_ppc.cpp	2017-09-27 11:38:58.726149949 -0300
@@ -201,6 +201,12 @@
     FLAG_SET_DEFAULT(UseSHA512Intrinsics, false);
   }
 
+  if (FLAG_IS_DEFAULT(UseMontgomeryMultiplyIntrinsic)) {
+    UseMontgomeryMultiplyIntrinsic = true;
+  }
+  if (FLAG_IS_DEFAULT(UseMontgomerySquareIntrinsic)) {
+    UseMontgomerySquareIntrinsic = true;
+  }
 }
 
 void VM_Version::print_features() {
