--- old/src/cpu/ppc/vm/macroAssembler_ppc.hpp	2017-08-29 17:29:16.184708721 -0300
+++ new/src/cpu/ppc/vm/macroAssembler_ppc.hpp	2017-08-29 17:29:16.056705713 -0300
@@ -812,6 +812,8 @@
                                Register yz_idx, Register idx, Register carry,
                                Register product_high, Register product,
                                Register carry2, Register tmp);
+  void muladd(Register out, Register in, Register offset, Register len, Register k,
+              Register tmp1, Register tmp2, Register carry);
   void multiply_to_len(Register x, Register xlen,
                        Register y, Register ylen,
                        Register z, Register zlen,
