--- old/src/cpu/ppc/vm/assembler_ppc.hpp	2017-08-25 12:46:07.111928014 -0300
+++ new/src/cpu/ppc/vm/assembler_ppc.hpp	2017-08-25 12:46:06.991928014 -0300
@@ -2174,7 +2174,8 @@
   inline void vsbox(       VectorRegister d, VectorRegister a);
 
   // SHA (introduced with Power 8)
-  // Not yet implemented.
+  inline void vshasigmad(VectorRegister d, VectorRegister a, bool st, int six);
+  inline void vshasigmaw(VectorRegister d, VectorRegister a, bool st, int six);
 
   // Vector Binary Polynomial Multiplication (introduced with Power 8)
   inline void vpmsumb(  VectorRegister d, VectorRegister a, VectorRegister b);
