--- old/test/compiler/intrinsics/sha/cli/testcases/GenericTestCaseForOtherCPU.java	2017-09-12 11:29:56.440464117 -0300
+++ new/test/compiler/intrinsics/sha/cli/testcases/GenericTestCaseForOtherCPU.java	2017-09-12 11:29:56.320461267 -0300
@@ -42,7 +42,8 @@
                               new OrPredicate(Platform::isAArch64,
                               new OrPredicate(Platform::isS390x,
                               new OrPredicate(Platform::isSparc,
-                              new OrPredicate(Platform::isX64, Platform::isX86))))));
+                              new OrPredicate(Platform::isPPC,
+                              new OrPredicate(Platform::isX64, Platform::isX86)))))));
     }
 
     @Override
