--- old/src/cpu/ppc/vm/assembler_ppc.hpp	2017-09-01 13:55:45.761938506 -0300
+++ new/src/cpu/ppc/vm/assembler_ppc.hpp	2017-09-01 13:55:44.665912757 -0300
@@ -1299,6 +1299,7 @@
   inline void li(   Register d, int si16);
   inline void lis(  Register d, int si16);
   inline void addir(Register d, int si16, Register a);
+  inline void subi( Register d, Register a, int si16);
 
   static bool is_addi(int x) {
      return ADDI_OPCODE == (x & ADDI_OPCODE_MASK);
